utils/logging/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+//! The second one, client-side by implementing check against Record Level field.
+//! Each module may have its own Logger or share one.
+
+use core::{cmp, fmt};
+
+static LOG_LEVEL_NAMES: [&str; 6] = ["GOSSIP", "TRACE", "DEBUG", "INFO", "WARN", "ERROR"];
+
+/// An enum representing the available verbosity levels of the logger.
+#[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
+pub enum Level {
+    /// Designates extremely verbose information, including gossip-induced messages
+    Gossip,
+    /// Designates very low priority, often extremely verbose, information
+    Trace,
+    /// Designates lower priority information
+    Debug,
+    /// Designates useful information
+    Info,
+    /// Designates hazardous situations
+    Warn,
+    /// Designates very serious errors
+    Error,
+}
+
+impl PartialOrd for Level {
+    #[inline]
+    fn partial_cmp(&self, other: &Level) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+
+    #[inline]
+    fn lt(&self, other: &Level) -> bool {
+        (*self as usize) < *other as usize
+    }
+
+    #[inline]
+    fn le(&self, other: &Level) -> bool {
+        *self as usize <= *other as usize
+    }
+
+    #[inline]
+    fn gt(&self, other: &Level) -> bool {
+        *self as usize > *other as usize
+    }
+
+    #[inline]
+    fn ge(&self, other: &Level) -> bool {
+        *self as usize >= *other as usize
+    }
+}
+
+impl Ord for Level {
+    #[inline]
+    fn cmp(&self, other: &Level) -> cmp::Ordering {
+        (*self as usize).cmp(&(*other as usize))
+    }
+}
+
+impl fmt::Display for Level {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.pad(LOG_LEVEL_NAMES[*self as usize])
+    }
+}
+
+impl Level {
+    /// Returns the most verbose logging level.
+    #[inline]
+    pub fn max() -> Level {
+        Level::Gossip
+    }
+}
+
+/// A Record, unit of logging output with Metadata to enable filtering
+/// Module_path, file, line to inform on log's source
+#[derive(Clone, Debug)]
+pub struct Record<'a> {
+    /// The verbosity level of the message.
+    pub level: Level,
+    /// The message body.
+    pub args: fmt::Arguments<'a>,
+    /// The module path of the message.
+    pub module_path: &'static str,
+    /// The source file containing the message.
+    pub file: &'static str,
+    /// The line containing the message.
+    pub line: u32,
+}
+
+impl<'a> Record<'a> {
+    /// Returns a new Record.
+    #[inline]
+    pub fn new(
+        level: Level,
+        args: fmt::Arguments<'a>,
+        module_path: &'static str,
+        file: &'static str,
+        line: u32,
+    ) -> Record<'a> {
+        Record {
+            level,
+            args,
+            module_path,
+            file,
+            line,
+        }
+    }
+}
+
+/// A trait encapsulating the operations required of a logger
+pub trait Logger {
+    /// Logs the `Record`
+    fn log(&self, record: &Record);
```

Comment:
I thought that we wanted an async logger @rrybarczyk ? If that is the case this function should be async or we should have 2 function one async and one not.

Reply:
Yes - it's true that we want an async logger. This is really a POC to get approval on the design. I'll be writing an async logger for the pool role next.

Reply:
Ok perfect, I like the POC

Reply:
Further - you can't currently use `async` in traits - I could pull in [async-trait crate from crates.io](https://github.com/dtolnay/async-trait) but in the rust docs it says "Note that using these trait methods will result in a heap allocation per-function-call. This is not a significant cost for the vast majority of applications, but should be considered when deciding whether to use this functionality in the public API of a low-level function that is expected to be called millions of times a second."

I feel like the `log` method is probably a low level function and might be called frequently.

Reply:
yes but if we want to write log to a db for example I guess that we want it to be async. If we do not want to be async we could implement the logger in a way that log just save the Record in internal state, and just spwan a task that read the internal state and write the new records where we want. We should have at least one StandardLogger for tokio and one for async_std

Reply:
utils is not part of core IMO

Reply:
Well, since we don't have a core yet I suppose that can be moved when core gets created?

Reply:
I think that everything under ./protocol/ is core ./utils/ are not part of core in fact nothing under ./protocols depend on ./utils/

Reply:
There are lots of `buffer_sv2` dependencies in protocol/v2 - codec-sv2, framing-sv2, noise-sv2, binary-sv2, no-serde-sv2. What am I missing?

Reply:
yep but it should always be optional, if not we should move it out